docs/architecture/adr-038-state-listening.md

```diff
@@ -182,349 +139,606 @@ func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
         store = tracekv.NewStore(store, rs.traceWriter, rs.traceContext)
     }
     if rs.ListeningEnabled(key) {
-        store = listenkv.NewStore(key, store, rs.listeners[key])
+        store = listenkv.NewStore(store, key, rs.listeners[key])
     }
 
     return store
 }
 ```
 
-We will also adjust the `cachemulti` constructor methods and the `rootmulti` `CacheMultiStore` method to forward the listeners
-to and enable listening in the cache layer.
+We will implement `AddListeners` to manage KVStore listeners internally and implement `PopStateCache`
+for a means of retrieving the current state.
+
+```go
+// AddListeners adds state change listener for a specific KVStore
+func (rs *Store) AddListeners(keys []types.StoreKey) {
+	listener := types.NewMemoryListener()
+	for i := range keys {
+		rs.listeners[keys[i]] = listener
+	}
+}
+```
+
+```go
+func (rs *Store) PopStateCache() []types.StoreKVPair {
+	var cache []types.StoreKVPair
+	for _, ls := range rs.listeners {
+		cache = append(cache, ls.PopStateCache()...)
+	}
+	return cache
+}
+```
+
+We will also adjust the `rootmulti` `CacheMultiStore` and `CacheMultiStoreWithVersion` methods to enable listening in
+the cache layer.
 
 ```go
 func (rs *Store) CacheMultiStore() types.CacheMultiStore {
     stores := make(map[types.StoreKey]types.CacheWrapper)
     for k, v := range rs.stores {
-        stores[k] = v
+        store := v.(types.KVStore)
```

Comment:
Didn't https://github.com/cosmos/cosmos-sdk/pull/13516 indicate that listening in the cache layer is generally not workable?

Reply:
Yes, it addresses the issue by moving listening to the `rootmulti.Store`. `rootmulti.Store` implements `CommitMultiStore` interface (an extension of `MultiStore`) . These two methods are how we will enable listening in `rootmulti`.

>We will also adjust the `rootmulti` `CacheMultiStore` and `CacheMultiStoreWithVersion` methods to enable listening in the cache layer.

Reply:
Why are we modifying the cache-MS to begin with? Personally I want to remove this type and abstraction all together. It's really not necessary.

Reply:
are you talking about removing  the `CacheMultiStore()` method from `CommitMultiStore`?